File: backend/client_workspace.py
"""
Block 9.0 — Client Workspace (READ-ONLY, AGGREGATION ONLY)

One endpoint: GET /api/client/project/{project_id}/workspace

Purpose (1 screen = 1 question):
    "What's happening with my project right now?"

Returns:
    - project header          (id, title)
    - status + cause + 1-line explanation
    - KPI summary             (revenue / cost / paid / profit)
    - modules                 (status / progress / developer / cost_status)

Intentionally does NOT return:
    - system_actions    → lives in /api/client/operator (that's its job)
    - opportunities     → lives in /api/client/operator/opportunities
    - pending_decisions → lives on individual module endpoints

No new business logic. Reuses shared helpers from client_operator
(_risk_state, _cost_status) — admin & client can never diverge on risk.
No new collections. Reads db.projects, db.modules, db.payouts, db.users.
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException

from client_operator import _cost_status, _risk_state

router = APIRouter(prefix="/api", tags=["client-workspace"])


def _explain(risk: str, over_count: int, warn_count: int) -> str:
    if risk == "blocked":
        return "System paused parts of the project to protect your budget."
    if risk == "at_risk":
        if over_count:
            return "One or more modules went over budget — review and rebalance."
        return "System is watching cost pressure across modules."
    if risk == "watch":
        return "System is monitoring performance and cost."
    return "Everything is running smoothly."


def _cause(risk: str, paused_sys: int, over_count: int, warn_count: int) -> Optional[str]:
    """Short 2–4 words for the header pill: 'Blocked · <cause>'."""
    if risk == "blocked":
        if over_count and paused_sys:
            return "margin risk"
        if over_count:
            return "over budget"
        if paused_sys:
            return "paused by system"
        return "margin risk"
    if risk == "at_risk":
        return "over budget" if over_count else "cost pressure"
    if risk == "watch":
        return "near limits" if warn_count else "monitoring"
    return None  # healthy → no subtitle


def _status_label(risk: str) -> str:
    return {
        "healthy": "Healthy",
        "watch": "Watch",
        "at_risk": "At risk",
        "blocked": "Blocked",
    }.get(risk, "Unknown")


# Sort priority for module list:
#   1. paused by system  (system intervened — most urgent)
#   2. over_budget
#   3. warning (near limit)
#   4. active (in_progress / review / pending)
#   5. paused (manual)
#   6. done / completed
def _module_sort_key(m: Dict[str, Any]) -> int:
    if m.get("paused_by_system"):
        return 0
    cs = m.get("cost_status")
    st = m.get("status")
    if cs == "over_budget":
        return 1
    if cs == "warning":
        return 2
    if st in ("in_progress", "review", "pending"):
        return 3
    if st == "paused":
        return 4
    return 5


def init_router(db, get_current_user_dep):

    @router.get("/client/project/{project_id}/workspace")
    async def client_workspace(project_id: str,
                               user=Depends(get_current_user_dep)) -> Dict[str, Any]:
        client_id = user.user_id if hasattr(user, "user_id") else user["user_id"]

        # 1. Project (auth scope)
        project = await db.projects.find_one(
            {"project_id": project_id, "client_id": client_id}, {"_id": 0},
        )
        if not project:
            raise HTTPException(status_code=404, detail="Project not found")

        # 2. Modules
        modules = await db.modules.find(
            {"project_id": project_id}, {"_id": 0},
        ).to_list(1000)
        module_ids = [m["module_id"] for m in modules]

        # 3. Payouts (earned = approved+paid, paid = paid) — same rule as Operator/Costs
        payouts_by_mod: Dict[str, List[Dict[str, Any]]] = {}
        if module_ids:
            all_p = await db.payouts.find(
                {"module_id": {"$in": module_ids}}, {"_id": 0},
            ).to_list(5000)
            for po in all_p:
                payouts_by_mod.setdefault(po["module_id"], []).append(po)

        # 4. Dev assignments → display names
        dev_ids = list({m.get("assigned_to") for m in modules if m.get("assigned_to")})
        dev_name_by_id: Dict[str, str] = {}
        if dev_ids:
            users = await db.users.find(
                {"user_id": {"$in": dev_ids}},
                {"_id": 0, "user_id": 1, "name": 1, "email": 1},
            ).to_list(500)
            for u in users:
                dev_name_by_id[u["user_id"]] = u.get("name") or u.get("email") or "Developer"

        # 5. Enrich modules (totals + risk rollup)
        tot_revenue = tot_cost = tot_paid = tot_earned = 0.0
        enriched_modules: List[Dict[str, Any]] = []
        risk_input: List[Dict[str, Any]] = []   # lean copy for _risk_state()

        for m in modules:
            mid = m["module_id"]
            payouts = payouts_by_mod.get(mid, [])
            earned = sum(float(x.get("amount") or 0) for x in payouts
                         if x.get("status") in ("approved", "paid"))
            paid = sum(float(x.get("amount") or 0) for x in payouts
                       if x.get("status") == "paid")
            revenue = float(m.get("final_price") or m.get("price") or 0)
            cost = float(m.get("base_price") or revenue)
            progress_ratio = (earned / cost) if cost > 0 else 0.0
            cost_status = _cost_status(earned, cost)
            status = m.get("status") or "pending"
            paused_by_system = (status == "paused" and m.get("paused_by") == "guardian")

            dev_id = m.get("assigned_to")
            dev_label = dev_name_by_id.get(dev_id or "", "") if dev_id else ""

            enriched_modules.append({
                "module_id": mid,
                "module_title": m.get("title") or "",
                "status": status,
                "paused_by_system": paused_by_system,
                "progress": round(min(progress_ratio, 1.0), 2),
                "progress_pct": int(min(progress_ratio, 1.0) * 100),
                "price": round(revenue, 2),
                "cost": round(cost, 2),
                "earned": round(earned, 2),
                "paid": round(paid, 2),
                "cost_status": cost_status,
                "developer_id": dev_id,
                "developer_name": dev_label,
            })
            risk_input.append({
                "status": status,
                "paused_by": m.get("paused_by"),
                "cost_status": cost_status,
            })

            tot_revenue += revenue
            tot_cost += cost
            tot_earned += earned
            tot_paid += paid

        # 6. Rollup
        risk = _risk_state(risk_input)
        over_count = sum(1 for m in enriched_modules if m["cost_status"] == "over_budget")
        warn_count = sum(1 for m in enriched_modules if m["cost_status"] == "warning")
        paused_sys = sum(1 for m in enriched_modules if m["paused_by_system"])
        active_count = sum(1 for m in enriched_modules
                           if m["status"] in ("pending", "in_progress", "review"))

        # BD-15 (slice #3) — additive module status counts.
        # Authority promotion per I-06: ≥2 frontend consumers (mobile project-
        # detail counters + mobile project-list inline counters) were doing
        # the same `.filter().length` synthesis client-side; web cabinet already
        # proves the same shape is server-computable via /full's workspace.
        # Narrow, additive, explicit. NOT a universal stats object.
        sc_in_progress = sum(1 for m in enriched_modules if m["status"] == "in_progress")
        sc_review = sum(1 for m in enriched_modules if m["status"] == "review")
        sc_done = sum(1 for m in enriched_modules
                      if m["status"] in ("done", "completed"))
        sc_paused = sum(1 for m in enriched_modules if m["status"] == "paused")

        # 7. Sort modules by priority (system-paused → over-budget → warning → active → …)
        enriched_modules.sort(key=_module_sort_key)

        # 8. Latest system action — the operator/guardian audit trail.
        # Surfacing the most recent line on the hero turns the project from
        # "managed by you" to "managed by the system, you supervise".
        last_action = await db.auto_actions.find_one(
            {"project_id": project_id},
            {"_id": 0, "type": 1, "impact": 1, "reason": 1, "created_at": 1},
            sort=[("created_at", -1)],
        )
        system_action = None
        if last_action:
            system_action = {
                "label": last_action.get("impact") or last_action.get("reason") or "System acted",
                "type": last_action.get("type"),
                "at": last_action.get("created_at"),
            }

        # Deposit block — surfaces 10% deposit required at conversion
        # (visitor → register → claim). Frontend renders a Pay-deposit card on
        # the project screen when `required: True`. Additive: legacy projects
        # without a deposit_amount/awaiting_deposit status get `required:False`.
        proj_status = project.get("status") or ""
        deposit_required = (
            proj_status == "awaiting_deposit"
            and not project.get("deposit_paid", False)
            and float(project.get("deposit_amount") or 0) > 0
        )
        deposit_block = {
            "required": bool(deposit_required),
            "paid": bool(project.get("deposit_paid", False)),
            "amount": float(project.get("deposit_amount") or 0),
            "final_price": float(project.get("final_price") or 0),
            "project_status": proj_status or None,
        }

        return {
            "project": {
                "project_id": project_id,
                "project_title": project.get("name") or project.get("title") or "",
                "status": proj_status or None,
                "created_at": project.get("created_at"),
            },
            "deposit": deposit_block,
            "summary": {
                "revenue": round(tot_revenue, 2),
                "cost": round(tot_cost, 2),
                "earned": round(tot_earned, 2),
                "paid": round(tot_paid, 2),
                "profit": round(tot_revenue - tot_cost, 2),
                "active_modules": active_count,
                "total_modules": len(enriched_modules),
                "over_budget_count": over_count,
                "warning_count": warn_count,
                "paused_by_system_count": paused_sys,
            },
            # BD-15 — module status counts (slice #3 promotion, I-06 satisfied).
            # Authoritative aggregation: frontend MUST NOT recompute.
            "status_counts": {
                "in_progress": sc_in_progress,
                "review": sc_review,
                "done": sc_done,
                "paused": sc_paused,
                "total": len(enriched_modules),
            },
            "status": risk,                 # healthy | watch | at_risk | blocked
            "status_label": _status_label(risk),
            "cause": _cause(risk, paused_sys, over_count, warn_count),
            "explanation": _explain(risk, over_count, warn_count),
            "system_action": system_action,
            "modules": enriched_modules,
            "generated_at": datetime.now(timezone.utc).isoformat(),
        }

    return router

File: web/src/pages/ClientWorkspace.js
/**
 * Block 9.0 — Client Workspace  (GET /api/client/project/:id/workspace)
 *
 * 1 screen = 1 question: "What is happening with my project right now?"
 *
 * Shows ONLY:
 *   - header     (project + status + 1-line cause)
 *   - KPI        (profit · revenue · cost · paid)
 *   - modules    (status · progress · developer · cost badge)
 *
 * Does NOT show (by design, to avoid mixing concerns):
 *   - system actions    → /client/operator
 *   - opportunities     → /client/opportunities (future)
 *   - earnings breakdown → /client/economics (future rename of /client/costs)
 */
import { useEffect, useState } from "react";
import { runtime } from '@/runtime';
import { ApiError } from '@/runtime-client';
import { useParams, Link } from "react-router-dom";
import { API } from "@/App";

const money = (n) =>
  `$${Number(n || 0).toLocaleString(undefined, { maximumFractionDigits: 0 })}`;

const STATUS = {
  healthy: { label: "Healthy", dot: "bg-emerald-400", border: "border-emerald-500/30", glow: "" },
  watch:   { label: "Watch",   dot: "bg-signal",     border: "border-signal/30",     glow: "shadow-signal/10"     },
  at_risk: { label: "At risk", dot: "bg-amber-400",   border: "border-amber-500/30",   glow: ""   },
  blocked: { label: "Blocked", dot: "bg-rose-400",    border: "border-rose-500/30",    glow: ""    },
};

const MOD_BADGE = {
  under_control: { label: "Under control", cls: "text-emerald-300 border-emerald-500/30 bg-emerald-500/10" },
  warning:       { label: "Near limit",    cls: "text-amber-300 border-amber-500/30 bg-amber-500/10"       },
  over_budget:   { label: "Over budget",   cls: "text-rose-300 border-rose-500/30 bg-rose-500/10"          },
};

export default function ClientWorkspace() {
  const { projectId } = useParams();
  const [data, setData] = useState(null);
  const [err, setErr] = useState(null);

  useEffect(() => {
    let active = true;
    setData(null);
    setErr(null);
    runtime
      .get(`/api/client/project/${projectId}/workspace`)
      .then((r) => active && setData(r.data))
      .catch((e) => active && setErr(e instanceof ApiError ? (e.hint || e.message) : (e?.response?.data?.detail || e.message)));
    return () => { active = false; };
  }, [projectId]);

  if (err)   return <div className="min-h-screen bg-[var(--t-bg)] text-rose-400 p-8">Failed: {String(err)}</div>;
  if (!data) return <div className="min-h-screen bg-[var(--t-bg)] text-muted-foreground p-8">Loading workspace…</div>;

  const status = STATUS[data.status] || STATUS.healthy;
  const profitClass = data.summary.profit >= 0 ? "text-emerald-400" : "text-rose-400";

  return (
    <div className="min-h-screen bg-[var(--t-bg)] text-white">
      <div className="max-w-4xl mx-auto px-6 py-10">

        {/* Breadcrumb */}
        <div className="text-muted-foreground text-xs mb-4 flex items-center gap-2">
          <Link to="/client/operator" className="hover:text-muted-foreground">Operator</Link>
          <span>/</span>
          <span className="text-muted-foreground">{data.project.project_title || "Project"}</span>
        </div>

        {/* Header */}
        <div className={`rounded-2xl border ${status.border} bg-white/[0.03] p-6 shadow-xl ${status.glow}`}>
          <div className="flex flex-wrap items-start justify-between gap-4">
            <div className="min-w-0">
              <h1 className="text-2xl font-bold">{data.project.project_title || "Untitled project"}</h1>
              <div className="mt-1 text-muted-foreground text-sm">{data.explanation}</div>
            </div>
            <span className="inline-flex items-center gap-2 text-xs px-2.5 py-1 rounded-md border border-border bg-black/40">
              <span className={`w-1.5 h-1.5 rounded-full ${status.dot}`} />
              {data.status_label}
              {data.cause && <span className="text-muted-foreground">· {data.cause}</span>}
            </span>
          </div>

          {/* KPI */}
          <div className="mt-5 grid grid-cols-2 md:grid-cols-4 gap-3">
            <MiniStat label="Profit"  value={money(data.summary.profit)}  color={profitClass} />
            <MiniStat label="Revenue" value={money(data.summary.revenue)} />
            <MiniStat label="Cost"    value={money(data.summary.cost)}    />
            <MiniStat label="Paid"    value={money(data.summary.paid)}    />
          </div>

          <div className="mt-3 text-muted-foreground text-xs">
            {data.summary.active_modules} active · {data.summary.total_modules} total
            {data.summary.over_budget_count > 0 && (
              <span className="ml-2 text-rose-400">· {data.summary.over_budget_count} over budget</span>
            )}
            {data.summary.paused_by_system_count > 0 && (
              <span className="ml-2 text-amber-400">· {data.summary.paused_by_system_count} paused by system</span>
            )}
          </div>
        </div>

        {/* Modules (full-width, no sidebar — this is a single-purpose screen) */}
        <div className="mt-6">
          <SectionTitle>Modules</SectionTitle>
          {data.modules.length === 0 && (
            <div className="text-muted-foreground text-sm">No modules yet.</div>
          )}
          <div className="space-y-2">
            {data.modules.map((m) => {
              const badge = MOD_BADGE[m.cost_status] || MOD_BADGE.under_control;
              return (
                <div key={m.module_id} className="rounded-lg border border-border bg-white/[0.02] p-3">
                  <div className="flex items-center justify-between gap-3">
                    <div className="min-w-0">
                      <div className="font-medium truncate">{m.module_title}</div>
                      <div className="text-muted-foreground text-xs mt-0.5">
                        {m.developer_name ? `Dev: ${m.developer_name}` : "Unassigned"}
                        {" · "}{m.status}{m.paused_by_system ? " (by system)" : ""}
                      </div>
                    </div>
                    <span className={`text-xs px-2 py-1 rounded-md border ${badge.cls}`}>
                      {badge.label}
                    </span>
                  </div>
                  <div className="mt-2 flex items-center gap-2">
                    <div className="flex-1 h-1.5 bg-muted rounded-full overflow-hidden">
                      <div
                        className={`h-full ${
                          m.cost_status === "over_budget" ? "bg-rose-400" :
                          m.cost_status === "warning"     ? "bg-amber-400" : "bg-emerald-400"
                        }`}
                        style={{ width: `${m.progress_pct}%` }}
                      />
                    </div>
                    <div className="text-xs text-muted-foreground w-10 text-right tabular-nums">
                      {m.progress_pct}%
                    </div>
                  </div>
                  <div className="mt-1 text-muted-foreground text-xs">
                    Cost {money(m.cost)} · Earned {money(m.earned)}
                  </div>
                </div>
              );
            })}
          </div>
        </div>

        {/* Deep links — each answers its own question */}
        <div className="mt-8 pt-4 text-xs space-y-1 text-muted-foreground border-t border-border">
          <div className="uppercase tracking-wider text-muted-foreground mb-1">Go deeper</div>
          <Link to="/client/operator" className="block hover:text-muted-foreground">Operator — what the system did · what you can do</Link>
          <Link to="/client/costs" className="block hover:text-muted-foreground">Economics — where the money is going</Link>
        </div>
      </div>
    </div>
  );
}

function MiniStat({ label, value, color }) {
  return (
    <div className="rounded-lg border border-border bg-black/30 px-3 py-2">
      <div className="text-[10px] uppercase tracking-wider text-muted-foreground">{label}</div>
      <div className={`mt-0.5 text-lg font-semibold tabular-nums ${color || "text-white"}`}>
        {value}
      </div>
    </div>
  );
}

function SectionTitle({ children }) {
  return (
    <div className="text-muted-foreground text-xs uppercase tracking-wider mb-2">{children}</div>
  );
}
